2021-12-27-getting-started-with-purenix.md

```diff
@@ -0,0 +1,217 @@
+
+```purescript
+module Main where
+
+import Prelude
+
+type FilePath = String
+
+subdirectory :: FilePath -> FilePath -> FilePath
+subdirectory p1 p2 = p1 <> "/" <> p2
+```
+
+The PureScript compiler compiles code like this to JavaScript.  PureScript ends
+up being a Haskell-like language that you can use to write frontend code.
+
+One of the main build tools for PureScript is
+[Spago](https://github.com/purescript/spago).  Spago is somewhat similar to
+Haskell's `stack` or `cabal`, Rust's `cargo`, etc.
+
+One interesting feature of the PureScript compiler and Spago is that there is
+good support for switching out the backend code generator.  The code generator
+built into the main PureScript compiler is for JavaScript, but there are
+alternative code generators for other languages.  PureNix is an alternative
+backend code generator for Nix.
+
+## PureNix
+
+PureNix works together with PureScript and Spago to compile PureScript code
+into Nix code.  Here's a small example of PureScript code, and the Nix code it
+gets compiled to:
+
+```purescript
+module Main where
+
+greeting :: String
+greeting = "Hello, world!"
+
+identity :: forall a. a -> a
+identity a = a
+
+data Maybe a = Nothing | Just a
+
+fromMaybe :: forall a. a -> Maybe a -> a
+fromMaybe a Nothing = a
+fromMaybe _ (Just a) = a
+```
+
+PureNix compiles this PureScript code to the following Nix code[^1]:
+
+```nix
+let
+  greeting = "Hello, world!";
+  Nothing = {__tag = "Nothing";};
+  Just = value0:
+    { __tag = "Just";
+      __field0 = value0;
+    };
+  fromMaybe = v: v1:
+    let
+      __pattern0 = __fail: if v1.__tag == "Nothing" then let a = v; in a else __fail;
+      __pattern1 = __fail: if v1.__tag == "Just" then let a = v1.__field0; in a else __fail;
+      __patternFail = builtins.throw "Pattern match failure in src/Main.purs at 11:1 - 11:41";
+    in
+      __pattern0 (__pattern1 __patternFail);
+  identity = v: v;
+in
+  {inherit greeting Nothing Just fromMaybe;}
+```
+
+Except for the pattern matching in `fromMaybe`, this should look pretty
+straight-forward.  You can see the PureScript string `greeting` becomes a
+simple Nix string.  `identity` has become a Nix function.  The constructors for
+`Maybe` (`Nothing` and `Just`) have become Nix functions[^2].  `fromMaybe` has
+become a Nix function.  If you work through the `__patternX` continuations in
+`fromMaybe`, you can see how they
+[emulate the pattern matching](https://jonascarpay.com/posts/2021-11-08-nix-adts.html)
+in the PureScript `fromMaybe` function.
+
+Calling functions written in PureScript from Nix is quite fun.  It feels quite
+similar to working in the Haskell or PureScript REPL.  Here is an example of
+using the above functions from the Nix REPL:
+
+```console
+$ nix repl ./output/Main/default.nix
+nix-repl> identity greeting
+"Hello, world!"
+nix-repl> fromMaybe "bye" (Just "hello")
+"hello"
+nix-repl> fromMaybe "bye" Nothing
+"bye"
+```
```

Comment:
I keep being surprised by how nice this works out. Maybe we should make a screencast out of this or something

Reply:
Yeah, we really should.  Aside from type-classes, it is super nice.